# Spam Detector

In [229]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score


## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [230]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

## Split the Data into Training and Testing Sets

In [231]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']
X = data.drop(columns=['spam'])


In [232]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()


spam
0    2788
1    1813
Name: count, dtype: int64

In [233]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)


In [234]:
# Display X_train DataFrame
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1947            0.00               0.00           0.00           0.0   
2159            0.08               0.17           0.17           0.0   
4223            0.00               0.00           1.85           0.0   
2624            0.00               0.00           0.00           0.0   
2743            0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1947            0.0            0.00               0.0                0.00   
2159            0.8            0.08               0.0                0.26   
4223            0.0            0.00               0.0                0.00   
2624            0.0            0.00               0.0                0.00   
2743            0.0            0.00               0.0                0.39   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1947              0.0            0.00  ...                   0.0        0.000   
2159              0.0            0.00  ...                   0.0        0.023   
4223              0.0            1.85  ...                   0.0        0.000   
2624              0.0            0.00  ...                   0.0        0.000   
2743              0.0            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1947        0.000          0.0        0.000        0.000         0.38   
2159        0.046          0.0        0.000        0.023         0.00   
4223        0.000          0.0        0.692        0.000         0.00   
2624        0.000          0.0        0.000        0.000         0.00   
2743        0.049          0.0        0.000        0.000         0.00   

      capital_run_length_average  capital_run_length_longest  \
1947                       2.307                           9   
2159                       2.658                          57   
4223                       1.727                           5   
2624                       1.000                           1   
2743                       2.000                          38   

      capital_run_length_total  
1947                        30  
2159                       436  
4223                        19  
2624                         8  
2743                       124  

[5 rows x 57 columns]

In [235]:
# Display the shape of the training and testing datasets
X_train.shape, X_test.shape, y_train.shape, y_test.shape


((3680, 57), (921, 57), (3680,), (921,))

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [236]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [237]:
# Fit the Standard Scaler with the training data
X_train_scaled = StandardScaler().fit_transform(X_train)
X_test_scaled = StandardScaler().fit_transform(X_test)



In [238]:
# Scale the training data
X_train_scaled = pd.DataFrame(X_train_scaled, columns=X_train.columns)
X_test_scaled = pd.DataFrame(X_test_scaled, columns=X_test.columns)
X_train_scaled.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0       -0.358493          -0.166316      -0.561452     -0.046223   
1       -0.086769          -0.034973      -0.223847     -0.046223   
2       -0.358493          -0.166316       3.112489     -0.046223   
3       -0.358493          -0.166316      -0.561452     -0.046223   
4       -0.358493          -0.166316      -0.561452     -0.046223   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0      -0.462132       -0.348265         -0.289743           -0.254338   
1       0.742377       -0.062859         -0.289743            0.380197   
2      -0.462132       -0.348265         -0.289743           -0.254338   
3      -0.462132       -0.348265         -0.289743           -0.254338   
4      -0.462132       -0.348265         -0.289743            0.697464   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0        -0.323326       -0.371634  ...              -0.10762    -0.156207   
1        -0.323326       -0.371634  ...              -0.10762    -0.070310   
2        -0.323326        2.411737  ...              -0.10762    -0.156207   
3        -0.323326       -0.371634  ...              -0.10762    -0.156207   
4        -0.323326       -0.371634  ...              -0.10762    -0.156207   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0    -0.484919    -0.151345    -0.320376    -0.311553     0.697670   
1    -0.324974    -0.151345    -0.320376    -0.216395    -0.104823   
2    -0.484919    -0.151345     0.463554    -0.311553    -0.104823   
3    -0.484919    -0.151345    -0.320376    -0.311553    -0.104823   
4    -0.314543    -0.151345    -0.320376    -0.311553    -0.104823   

   capital_run_length_average  capital_run_length_longest  \
0                   -0.108199                   -0.214503   
1                   -0.094012                    0.018188   
2                   -0.131641                   -0.233894   
3                   -0.161025                   -0.253285   
4                   -0.120607                   -0.073919   

   capital_run_length_total  
0                 -0.435240  
1                  0.260828  
2                 -0.454099  
3                 -0.472958  
4                 -0.274081  

[5 rows x 57 columns]

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [239]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
model_LR = LogisticRegression()
model_LR.fit(X_train_scaled, y_train)


LogisticRegression()

In [240]:
# Print the model score
print(f"Training Data Score: {model_LR.score(X_train_scaled, y_train)}")
print(f"Test Data Score: {model_LR.score(X_test_scaled, y_test)}")

Training Data Score: 0.9293478260869565
Test Data Score: 0.9261672095548317


In [241]:
# Make and save testing predictions with the saved logistic regression model using the test data
y_pred_LR = model_LR.predict(X_test_scaled)

# Review the predictions
y_pred_LR[:5]


array([0, 0, 1, 0, 0], dtype=int64)

In [242]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score_LR = accuracy_score(y_test, y_pred_LR)
print(f"Accuracy Score_LR: {acc_score_LR}")


Accuracy Score_LR: 0.9261672095548317


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [243]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
model_RF = RandomForestClassifier(n_estimators=128, random_state=78)
model_RF.fit(X_train_scaled, y_train)

RandomForestClassifier(n_estimators=128, random_state=78)

In [244]:
# Print the model score
print(f"Training Model Score: {model_RF.score(X_train_scaled, y_train)}")
print(f"Test Model Score: {model_RF.score(X_test_scaled, y_test)}")

Training Model Score: 0.9997282608695652
Test Model Score: 0.9500542888165038


In [245]:
# Make and save testing predictions with the saved logistic regression model using the test data
y_pred_RF = model_RF.predict(X_test_scaled)

# Review the predictions
y_pred_RF[:5]


array([0, 0, 1, 0, 0], dtype=int64)

In [246]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test, y_pred_RF)
print(f"Accuracy Score_RF: {acc_score}")


Accuracy Score_RF: 0.9500542888165038


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

Following are the scores for the Logistic Regression and Random Forest Classifier models:

- Logistic Regression Model - scores for Training data, Test data, and model accuracy scores, respectively: 
0.929, 0.926, and 0.926

- Random Forest Model - scores for Training data, Test data, and model accuracy scores, respectively: 
0.999, 0.950, and 0.950

From the scores above, the Random Forest Classifier model performed better than the Logistic Regression model.  It correctly predicted the spam instances 95% of the time (number of correct predictions (test data)/total number of predictions = 0.950), which was better than the Logistic Regression model at 0.926.   

In summary, the Random Forest Classification Model will detect spam email with 95% accuracy.